tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,253 @@
+use sharded_slab::{Guard, Slab};
+
+use super::stack::SpanStack;
+use crate::{
+    registry::{
+        extensions::{Extensions, ExtensionsInner, ExtensionsMut},
+        LookupSpan, SpanData,
+    },
+    sync::RwLock,
+};
+use std::{
+    cell::RefCell,
+    sync::atomic::{fence, AtomicUsize, Ordering},
+};
+use tracing_core::{
+    dispatcher,
+    field::FieldSet,
+    span::{self, Current, Id},
+    Event, Interest, Metadata, Subscriber,
+};
+
+/// A shared, reusable store for spans.
+///
+/// This registry is implemented using a [lock-free sharded slab][1], and is
+/// highly optimized for concurrent access.
+///
+/// [1]: https://docs.rs/crate/sharded-slab/0.0.5
+#[derive(Debug)]
+pub struct Registry {
+    spans: Slab<DataInner>,
+}
+
+/// Span data stored in a [`Registry`].
+///
+/// This definition  is intentionally kept small—only data pertaining to span
+/// relationships, span metadata, and active references are stored. Additional
+/// data, such as formatted fields, may be stored in the [extensions] typemap.
+///
+/// [`Registry`]: ../struct.Registry.html
+/// [extensions]: ../extensions/index.html
+#[derive(Debug)]
+pub struct Data<'a> {
+    inner: Guard<'a, DataInner>,
+}
+
+#[derive(Debug)]
+struct DataInner {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    ref_count: AtomicUsize,
+    pub(crate) extensions: RwLock<ExtensionsInner>,
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self { spans: Slab::new() }
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: DataInner) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, DataInner>> {
+        self.spans.get(id_to_idx(id))
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
```

Comment:
This is weird to me. Why is `Registry` a `Subscriber`??

Reply:
@jonhoo I added more API docs to `Registry` in 111d425c16c93e21b00371f00891bf5e74708db2, hopefully, they should make the answer to this question clearer. Let me know if there's anything still missing!